### PR TITLE
Restructure readers

### PR DESCRIPTION
**Pull request by wkerzendorf**
Very long PR to restructure the readers and sort them in individual files. A lot of cleanup that makes the files structure clearer and follows the other restructure process.

### CODE REVIEW

File: tardis/io/model/parse_density_configuration.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    time_explosion.
+
+    Parameters
+    ----------
+
+    density_configuration : tardis.io.config_reader.Configuration
+    v_middle : astropy.Quantity
+        middle of the velocity bins
+    time_explosion : astropy.Quantity
+        time of the explosion
```
Comment (jvshields) [inline tardis/io/model/parse_density_configuration.py:24]: Docstring should describe type and structure of objects returned.

File: tardis/io/model/readers/generic_readers.py [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+        names=("index", "velocity", "density"),
+        dtype=(int, float, float),
+    )
+    velocity = (data["velocity"] * u.km / u.s).to("cm/s")
+    mean_density = (data["density"] * u.Unit("g/cm^3"))[1:]
+
+    return time_of_model, velocity, mean_density
+
+
+def read_artis_density(fname):
```
Comment (jvshields) [inline tardis/io/model/readers/generic_readers.py:59]: Is there a reason the artis reader lives in the generic readers file? Would it be more clear to live in an artist reader file similar to cmfgen currently?

**Comment by andrewfullard** [comment]
I think this needs a rebase.

File: tardis/io/model/readers/generic_readers.py (hunk 2 above)
Comment (andrewfullard) [reply tardis/io/model/readers/generic_readers.py:59]: Agreed

File: tardis/io/model/readers/tests/test_arepo_parser.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def get_cone_csvy_model(tardis_ref_path):
 
 
 @pytest.fixture
-def get_full_csvy_model(tardis_ref_path):
+def get_full_csvy_model(tardis_ref_path, example_model_file_dir: Path):
```
Comment (andrewfullard) [inline tardis/io/model/readers/tests/test_arepo_parser.py:59]: Very inconsistent use of the type hinting here

File: tardis/io/model/readers/util.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+import numpy as np
+import pandas as pd
+from radioactivedecay import Nuclide
+from radioactivedecay.utils import Z_DICT, elem_to_Z
+
+
+def read_csv_isotope_abundances(
```
Comment (andrewfullard) [inline tardis/io/model/readers/util.py:7]: This is not *just* isotope abundances, correct?

File: tardis/io/model/readers/cmfgen.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    abundances.
+
+    WARNING : deprecated
+
+    fname : str
+        filename of the csv file
+    """
+
+    warnings.warn(
+        "The current CMFGEN model parser is deprecated", DeprecationWarning
```
Comment (andrewfullard) [inline tardis/io/model/readers/cmfgen.py:86]: Do we know why this is deprecated?

File: tardis/io/configuration/config_validator.py [hunk 6]
```diff
@@ -83,11 +83,13 @@ def is_quantity(checker, instance):
...
 def validate_dict(
-    config_dict, schemapath=config_schema_file, validator=DefaultDraft4Validator
+    config_dict,
+    schemapath=CONFIG_SCHEMA_FNAME,
+    validator=DefaultDraft4Validator,
 ):
     with open(schemapath) as f:
         schema = yaml.load(f, Loader=YAMLLoader)
-    schemaurl = "file://" + schemapath
+    schemaurl = f"file://{schemapath}"
```
Comment (andrewfullard) [inline tardis/io/configuration/config_validator.py:92]: I'm not sure if the format string is actually more readable than the concatenation here.

File: tardis/io/model/readers/util.py (hunk 4 above)
Comment (wkerzendorf) [reply tardis/io/model/readers/util.py:7]: yes

File: tardis/io/model/readers/cmfgen.py (hunk 5 above)
Comment (wkerzendorf) [reply tardis/io/model/readers/cmfgen.py:86]: no idea. There is a CMFGen reader somewhere else but I did not dig into it. I think there is a chance that this misinterprets CMFGen but I did not dig deeper.

File: tardis/io/configuration/config_validator.py (hunk 6 above)
Comment (wkerzendorf) [reply tardis/io/configuration/config_validator.py:92]: I'm not a fan of string additions and thing this is much cleaner - unless you strongly disagree

File: tardis/io/model/parse_density_configuration.py (hunk 1 above)
Comment (wkerzendorf) [reply tardis/io/model/parse_density_configuration.py:24]: I've added the type hints now but can also add a description

File: tardis/io/configuration/config_validator.py (hunk 6 above)
Comment (sonachitchyan) [reply tardis/io/configuration/config_validator.py:92]: Is there a lot of difference between the two methods??

File: tardis/io/model/readers/tests/test_arepo_parser.py [hunk 7]
```diff
@@ -1,16 +1,16 @@
...
+from pathlib import Path
+from tardis.io.model.readers import arepo
 import pytest
 import numpy as np
 import os
 import json
 
 
-DATA_PATH = os.path.join(tardis.__path__[0], "io", "tests", "data")
+# DATA_PATH = os.path.join(tardis.__path__[0], "io", "tests", "data")
```
Comment (andrewfullard) [inline tardis/io/model/readers/tests/test_arepo_parser.py:9]: Why not just delete?

Comment (wkerzendorf) [reply tardis/io/model/readers/tests/test_arepo_parser.py:9]: done!

File: tardis/io/configuration/config_validator.py (hunk 6 above)
Comment (wkerzendorf) [reply tardis/io/configuration/config_validator.py:92]: f-strings do automatic string conversion. I think also we try to use those more in python. I think it's a stylistic choice mostly.

File: tardis/io/model/readers/generic_readers.py (hunk 2 above)
Comment (wkerzendorf) [reply tardis/io/model/readers/generic_readers.py:61]: that should have been removed and moved to `artis.py`. It's not showing up in the diff but is definitely there. check it.

**Review by andrewfullard — DISMISSED** [review]

**Comment by wkerzendorf** [comment]
> _*beep* *bop*_ Hi human, I ran benchmarks as you asked comparing master ([a2af876](https://github.com/tardis-sn/tardis/commit/a2af8762cb24bfc78dbb746716419374d9e9c082)) and the latest commit ([2d030d2](https://github.com/tardis-sn/tardis/commit/2d030d24d42d8d47409319c2effd6a9b56fb8d50)). Here are the logs produced by ASV. Results can also be downloaded as artifacts [**here**](https://github.com/tardis-sn/tardis/actions/runs/5915000486?check_suite_focus=true). Significantly changed benchmarks:
> 
> All benchmarks:
> 
> ```diff
> All benchmarks:
> 
>      before           after         ratio
>    [a2af8762]       [2d030d24]
>    <release-2023.08.20~5>                 
>          failed           failed      n/a  benchmark_run_tardis.Benchmarkruntardis.time_run_tardis
> ```

@atharva-2001 something seems off

**Review by andrewfullard — APPROVED** [review]

**Review by sonachitchyan — APPROVED** [review]

**Pull request merged**